packages/std/src/math.rs

```diff
@@ -0,0 +1,255 @@
+/// That is Decimal9(1_000_000_000) == 1
+#[derive(Serialize, Deserialize, Copy, Clone, Debug, PartialEq, JsonSchema)]
+pub struct Decimal9(pub u64);
+
+pub const DECIMAL_FRACTIONAL: u64 = 1_000_000_000;
+
+impl Decimal9 {
+    pub fn one() -> Decimal9 {
+        Decimal9(DECIMAL_FRACTIONAL)
+    }
+
+    // convert integer % into Billionth units
+    pub fn percent(percent: u64) -> Decimal9 {
+        Decimal9(percent * 10_000_000)
+    }
+
+    // convert permille (1/1000) into Billionth units
+    pub fn permille(permille: u64) -> Decimal9 {
+        Decimal9(permille * 1_000_000)
+    }
+}
+
+//*** Uint128 ***/
+#[derive(Copy, Clone, Default, Debug, PartialEq, Eq, PartialOrd, Ord, JsonSchema)]
+pub struct Uint128(#[schemars(with = "String")] pub u128);
+
+impl Uint128 {
+    pub fn u128(&self) -> u128 {
+        self.0
+    }
+}
+
+impl From<u128> for Uint128 {
+    fn from(val: u128) -> Self {
+        Uint128(val)
+    }
+}
+
+impl From<u64> for Uint128 {
+    fn from(val: u64) -> Self {
+        Uint128(val.into())
+    }
+}
+
+impl TryFrom<&str> for Uint128 {
+    type Error = StdError;
+
+    fn try_from(val: &str) -> Result<Self, Self::Error> {
+        match val.parse::<u128>() {
+            Ok(u) => Ok(Uint128(u)),
+            Err(e) => Err(generic_err(format!("Parsing coin: {}", e))),
+        }
+    }
+}
+
+impl Into<String> for Uint128 {
+    fn into(self) -> String {
+        self.0.to_string()
+    }
+}
+
+impl Into<u128> for Uint128 {
+    fn into(self) -> u128 {
+        self.u128()
+    }
+}
+
+impl fmt::Display for Uint128 {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl ops::Add for Uint128 {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        Uint128(self.u128() + other.u128())
+    }
+}
+
+impl ops::AddAssign for Uint128 {
+    fn add_assign(&mut self, other: Self) {
+        self.0 += other.u128();
+    }
+}
+
+impl ops::Sub for Uint128 {
+    type Output = StdResult<Self>;
+
+    fn sub(self, other: Self) -> StdResult<Self> {
+        let (min, sub) = (self.u128(), other.u128());
+        if sub > min {
+            Err(underflow(min, sub))
+        } else {
+            Ok(Uint128(min - sub))
+        }
+    }
+}
+
+impl ops::Mul<Decimal9> for Uint128 {
+    type Output = Self;
+
+    fn mul(self, rhs: Decimal9) -> Self {
+        self.multiply_ratio(rhs.0.into(), DECIMAL_FRACTIONAL.into())
+    }
+}
+
+impl Uint128 {
+    /// returns self * num / denom
+    pub fn multiply_ratio(&self, num: Uint128, denom: Uint128) -> Uint128 {
+        // TODO: minimize rounding that takes place (using gcd algorithm)
+        let val = self.u128() * num.u128() / denom.u128();
+        Uint128::from(val)
+    }
+
+    /// Returns the ratio (self / denom) as Decimal9 fixed-point
+    pub fn calc_ratio(&self, denom: Uint128) -> Decimal9 {
+        // special case: 0/0 = 1.0
```

Comment:
Mathematically this is undefined. If there is no very very godd reasonto add such a special case, division by 0 should always error or panic

Reply:
Okay, then we need to handle that in contract logic. Feel free to update that to be more correct. I was just doing the simple divide calculation for prices and this was the easiest "fix".

If you look at the math (multiple and divide) I am doing in the contract, then you can make any adjustments to make this more correct. I was just trying to get it to compile and pass first tests... and know this one needs some fixes.